cmd/hubble/daemon_linux.go

```diff
@@ -66,6 +68,16 @@ var (
 				return nil, nil, fmt.Errorf("creating new controller-runtime manager: %w", err)
 			}
 
+			if err := ctrlManager.AddHealthzCheck("healthz", healthz.Ping); err != nil {
+				logger.Error("unable to set up healthz check", err)
+				os.Exit(1)
```

Comment:
`os.Exit` is a heavy hammer, to be used with extreme discretion. It bypasses all deferred functions in Go, which is the primary way that resource cleanup happens. It's actually slightly better to panic (but also do that as rarely as possible) since there's at least a chance of it getting `recover()`ed by something further up the call stack. **Always prefer to return errors** unless there's just really no way to do it. Save panics for "developer errors" where there's no way for the compiler to check something at build time and you want it to blow up loudly (think compiling a regexp at runtime that happens to be invalid).

For reference, I usually start most Go applications like this:

```go
package main

func main() {
       if err := run(); err != nil {
              fmt.Println(err)
              os.Exit(1) // this is the only place in the entire application this appears, and really only because I want the exit code to be 1 instead of 0.
       }
}

func run() error {
       if err := somethingBadCouldHappen(); err != nil {
              return fmt.Errorf("invoking something potentially bad: %w", err) 
       }
       return nil
}
```

Reply:
@timraymond  That makes sense, though in this particular case, I'm wondering if it can recover? If, for example the health check creation fails and we just return an error, the /healthz endpoint is never set up, we still have the probe that will try and GET from this endpoint. If that fails the pods will keep restarting and won't become available.

Reply:
@mereta Sounds good. It might be okay in this instance, but in my experience it's not even worth reasoning about. Logic can be sound at this one moment in time, but as changes happen in the codebase someone still needs to remember that panic / recover is broken here and their defers will not run (in any goroutine). In general, it creates surprising behavior in the language and suppresses vital debugging information if something goes wrong. Being extremely judicious with `os.Exit` as a general principle will keep an entire class of spooky bugs out of the Go you write (honestly, we should have a linter for this).

Reply:
> honestly, we should have a linter for this.

kinda https://github.com/microsoft/retina/pull/1092/files/df90b99316baf690bd09504517875b8a31bd2328..a200860d63a1bb65c815f0707e9b3e8ba65ecdfc

Reply:
Oooh, that one is fancy. I'm impressed the static analysis is good enough to tell what defers won't run.